ivadomed/utils.py

```diff
@@ -34,6 +34,91 @@ def get_results(self):
         return res_dict
 
 
+class Evaluation3DMetrics(object):
+
+    def __init__(self, fname_pred, fname_gt, params=None):
+        self.fname_pred = fname_pred
+        self.fname_gt = fname_gt
+
+        if not params is None:
+            pass
+
+        self.data_pred = self.get_data(self.fname_pred)
+        self.data_gt = self.get_data(self.fname_gt)
+
+        self.px, self.py, self.pz = self.get_pixdim(self.fname_pred)
+
+    def get_data(self, fname):
+        nib_im = nib.load(fname)
+        return nib_im.get_data()
+
+    def get_pixdim(self, fname):
+        nib_im = nib.load(fname)
+        px, py, pz = nib_im.header['pixdim'][1:4]
+        return px, py, pz
+
+    def get_vol(self, data):
+        vol = np.sum(data)
+        vol *= self.px * self.py * self.pz
+        return vol
+
+    def get_rvd(self):
+        """Relative volume difference."""
+        vol_gt = self.get_vol(self.data_gt)
+        vol_pred = self.get_vol(self.data_pred)
+
+        if vol_gt == 0.0:
+            return np.nan
+
+        rvd = (vol_gt-vol_pred)*100.
+        rvd /= vol_gt
+        return rvd
+
+    def get_avd(self):
+        """Absolute volume difference."""
+        return abs(self.get_rvd())
+
+    def get_dice(self):
+        return dice_score(self.data_gt, self.data_pred) * 100.
+
+    def get_precision(self):
```

Comment:
This is duplicate from medicaltorch metrics.py, the methods get_precision, gt_recall, get_specificity could be removed and the methods from medical torch could be called directly in get_all_metrics

Reply:
The reason I created a method `get_dice` is to allow the user to get only dice score as metric from this class (compared to `get_all_metrics`), sort of way to be more modular. But it could be modified. What do you think?

`get_precision`, `gt_recall`, `get_specificity` are not exactly the same than in `medicaltorch` --> return `np.nan` instead of `0` in case of null denominator. Does this make sense? Similar to `dice_score` issue.

Reply:
It could be an option, but if we want to do so, we should allow the option between only dice or all metrics in the config file no? But I am not sure of the necessity, if all metrics are returned which include dice, is it necessary to give the option? The only reason I see is if there is a significant difference in computing time to measure all metrics vs only dice (I didn't test this)

For nan vs 0, could this differentiation be beneficial for 2Dmetrics as well? If so metrics.py could be modified and we could avoid method duplication.

Reply:
Initially, it was one for the reason I created the `param` attribute. The class could also be called independently, as a debug function.
But I see what you mean, and agree, I will modify this.

`nan` vs `0` --> when we had this debate for the `dice_score` we thought that it was quite specific and we did not wanted to impose this to `medicaltorch`. I mean: in GM segmentation (Christian's project) all your samples are positive (ie 1 true voxel at least) --> so you do not meet the issue that we have where both pred and gt can be empty, and we do not necessary want to include these values.
Alternatively, I can modify the medical torch functions by adding a parameter `err_value` (or something like this) which returns 0 or nan depending on your case. Cool, thank you for feedback.

Reply:
Oh I see! Fair point!